### PR TITLE
Account for 64B DRAM read alignment on BH

### PR DESCRIPTION
**Pull request by abhullar-tt**
### Ticket
- https://github.com/tenstorrent/tt-metal/issues/8706
- issue closed because it was pushed and validated on `abhullar/bh-bringup`

### Problem description
- BH has 64B alignment for DRAM reads and 16B alignment for  DRAM writes but Metal does not track different wr/rd alignments

### What's changed
- Added L1/DRAM/PCIE read/write alignment constraints in noc/noc_parameters.h
- Set L1/DRAM/PCIE alignment to be max of rd/wr alignments 
- Use max of L1 and DRAM alignment for L1/DRAM allocators 
- Add accessor to buffer to get aligned page size/buffer size

### Checklist
- [x] [Post commit CI](https://github.com/tenstorrent/tt-metal/actions/runs/9622992990)
- [x] [Model regression CI](https://github.com/tenstorrent/tt-metal/actions/runs/9626837653) matches current regressions on main

### CODE REVIEW

**Review by pgkeller — APPROVED** [review]
so much better!
(eventually we should move the host side out of hostdevcommon, that can be done later as I have no idea where it would go at present)

**Comment by tt-aho** [comment]
We might want to keep all sharded buffers aligned to the highest alignment for now (dram alignment) so that there aren't any issues with read/write between dram, and the sharded l1 buffer. Example if we run interleaved to sharded with a dram buffer as input and a sharded l1 buffer as the output, the read address is a 32B aligned dram addr, but the write address is a 16B aligned l1 address, then this doesn't work right?

**Comment by abhullar-tt** [comment]
> We might want to keep all sharded buffers aligned to the highest alignment for now (dram alignment) so that there aren't any issues with read/write between dram, and the sharded l1 buffer. Example if we run interleaved to sharded with a dram buffer as input and a sharded l1 buffer as the output, the read address is a 32B aligned dram addr, but the write address is a 16B aligned l1 address, then this doesn't work right?

Yes, good catch. Will update that

File: tt_metal/impl/buffers/buffer.hpp [hunk 1]
```diff
@@ -223,6 +225,12 @@ class Buffer {
 
     uint64_t page_address(uint32_t bank_id, uint32_t page_index) const;
 
+    uint32_t alignment() const { return ALLOCATOR_ALIGNMENT; }
```
Comment (tarafdarTT) [inline tt_metal/impl/buffers/buffer.hpp:228]: FUCK YES thank you for these :). Let's get these in ASAP because I need this.

**Review by ntarafdar — APPROVED** [review]
Amazing! Thanks @abhullar-tt

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 2]
```diff
@@ -1834,7 +1832,7 @@ void HWCommandQueue::copy_into_user_space(
 
         if (dev_page_to_host_page_mapping.empty()) {
             void* contiguous_dst = (void*)(uint64_t(dst) + contig_dst_offset);
-            if ((page_size % ADDRESS_ALIGNMENT) == 0) {
+            if (page_size != padded_page_size) {
```
Comment (tt-aho) [inline tt_metal/impl/dispatch/command_queue.cpp:1835]: Shouldn't this be ==?

**Review by tt-aho — APPROVED** [review]

File: tt_metal/impl/dispatch/command_queue.cpp (hunk 2 above)
Comment (abhullar-tt) [reply tt_metal/impl/dispatch/command_queue.cpp:1835]: thanks for catching

File: tt_metal/hw/inc/debug/sanitize_noc.h [hunk 3]
```diff
@@ -129,19 +129,17 @@ uint32_t debug_sanitize_noc_addr(
 
     // Check noc addr, we save the alignment requirement from the noc src/dst because the L1 address
     // needs to match alignment.
-    uint32_t alignment_mask = NOC_L1_ALIGNMENT_BYTES - 1;  // Default alignment, only override in ceratin cases.
+    uint32_t alignment_mask = (dir == DEBUG_SANITIZE_NOC_READ ? NOC_L1_READ_ALIGNMENT_BYTES : NOC_L1_WRITE_ALIGNMENT_BYTES) - 1;  // Default alignment, only override in ceratin cases.
```
Comment (abhullar-tt) [inline tt_metal/hw/inc/debug/sanitize_noc.h:132]: @tt-dma can you review these changes?

Comment (tt-dma) [reply tt_metal/hw/inc/debug/sanitize_noc.h:132]: Looks good to me, can we just update the comments on line 135/141? They're out of date now that we have restrictions for writes as well

**Review by tt-dma — APPROVED** [review]

**Pull request merged**